### DISCUSSION TITLE
How can I get the blob url when I finished zip

### DISCUSSION THREAD

**Discussion by hugetiny** [gildas-lormeau/zip.js#277] [Q&A] [answered]
I console.log all the r,result,blob, can't find info like:
blob:http://localhost:8080/xxxxxxxxxxxx

```
					const blobWriter = new zip.BlobWriter("application/zip");
					const zipWriter = new zip.ZipWriter(blobWriter, {
						level: 9
					});
					const r = await zipWriter.add(fileName, new zip.BlobReader(inputBlob));
					const result = await zipWriter.close()
					const blob = await blobWriter.getData()
```

**Comment by gildas-lormeau** [comment]
You need to call `const blobURL = Object.createURL(blob)` (see https://developer.mozilla.org/docs/Web/API/URL/createObjectURL) after the code you posted and `blobURL` will contain the string you want.

**Reply by hugetiny** [reply]
I tried 
```
const reader = new zip.ZipReader(r).config.workerScripts.deflate
```

and got the url.

Is that accurate for the results？

**Reply by gildas-lormeau** [reply]
No, this is the URL of the web worker used to decompress the file. If you open it in a new tab, you'll see it's a script.

**Reply by hugetiny** [reply]
I find out that  result and blob are the same.
So which line should I delete?
```
const result = await zipWriter.close()
const blob = await blobWriter.getData()
```

**Reply by gildas-lormeau** [reply]
I confirm that both `zipWriter.close()` and `blobWriter.getData()` return the Blob object containing the zip file. In your example, you should write:
```js
const blob = await zipWriter.close();
```

**Reply by hugetiny** [reply]
```		
const blobURL = URL.createObjectURL(blob)
```
I got the url like :blob:http://localhost:8080/504a1506-4588-4397-83a2-34e181a8f915 

But there are no data in blobURL 
If I type the url in Chrome.Chrome said "ERR_FILE_NOT_FOUND"

**Reply by gildas-lormeau** [reply]
I don't know what your problem is. On my end, this code works as expected: https://plnkr.co/edit/H8XmyMSOre1dyz2l?open=lib%2Fscript.js&preview

**Comment by hugetiny** [comment, marked as answer]
I want to select a file, zip it ,and upload to server.
The code is like :
```
const inputBlob = "blob:http......."
const blobWriter = new zip.BlobWriter("application/zip");
const zipWriter = new zip.ZipWriter(blobWriter, {level: 9});
const r = await zipWriter.add(fileName, new zip.BlobReader(inputBlob));
//const result = await zipWriter.close()
const blob = await blobWriter.getData()
const blobURL = URL.createObjectURL(blob)
window.open(blobURL, "_blank");
```
And finally I got the file 1kb ~ 2kb that can't be read

**Reply by gildas-lormeau** [reply]
In your example, `inputBlob` **is not** a `Blob` object, it's a `string` containing a Blob URL. You can use `window.fetch` to convert it to a `Blob`, e.g.

```js
const fetchResponse = await fetch(inputBlob); // inputBlob contains the Blob URL
const blob = await fetchResponse.blob()); // blob contains the corresponding Blob object
const blobWriter = new zip.BlobWriter("application/zip");
const zipWriter = new zip.ZipWriter(blobWriter, { level: 9 });
await zipWriter.add(fileName, new zip.BlobReader(blob));
const outputBlob = await zipWriter.close();
const outputBlobURL = URL.createObjectURL(outputBlob);
window.open(outputBlobURL, "_blank");
````

**Reply by hugetiny** [reply]
Thank you！
I got the file in right size，but I still can't open it in windows explorer or other zip file program

**Reply by gildas-lormeau** [reply]
I don't know the cause of your issue. This example https://plnkr.co/edit/cKiEEThZqu6ikG4r?open=lib%2Fscript.js&preview works fine for me.

**Reply by hugetiny** [reply]
await zipWriter.close() is the right choice for the final outputBlobURL .Thanks a lot
```
// const blob = await blobWriter.getData()
const blob = await zipWriter.close()
```

**Reply by gildas-lormeau** [reply]
Indeed, you **must always** close the zip file.
